sandman/source/logger.h

```diff
@@ -1,79 +1,351 @@
+
+	static constexpr std::string_view kMissingFormatValue = "`null`";
+
+	// "Lower-level" format write function.
+	// 
+	// Base case for template recursion. 
+	void FormatWrite(std::string_view const formatString);
+
+	// "Lower-level" format write function.
+	// 
+	// Interpolates arguments into the format string
+	// as it writes all characters in the format string to
+	// the buffer.
+	// 
+	// `%$` interpolates an argument into the format string.
+	// `%%` writes a literal percent sign `%` character.
+	//
+	// If interpolation is denoted with `%$` but there are no more arguments left
+	// to interpolate, `::Logger::kMissingFormatValue` is written instead.
+	// 
+	// The arguments are interpolated into the string in the order
+	// that they are passed into this function from left to right.
+	//
+	// If there are more arguments than there are interpolation indicators `%$`
+	// in the format string, the extra arguments are ignored.
+	template <typename T, typename... ParamsT>
+	void FormatWrite(std::string_view formatString, T&& firstArg, ParamsT&&... args);
+
+	// Object wrapper for formatting.
+	//
+	// Instances of this class can be passed to `Logger::WriteLine`.
+	// If so, the arguments will be interpolated into the
+	// format string and written to the buffer as if
+	// `Logger::FormatWrite` was called with the arguments
+	// that the `Format` instance was constructed with.
+	template <typename... ObjectsT>
+	struct Format
+	{
+		std::tuple<ObjectsT...> m_Objects;
+		std::string_view m_FormatString;
+
+		// Constructor.
+		//
+		// `%$` interpolates an argument into the format string.
+		// `%%` writes a literal percent sign `%` character.
+		//
+		// If interpolation is denoted with `%$` but there are no more arguments
+		// to interpolate, `::Logger::kMissingFormatValue` is written instead.
+		//
+		// The arguments are interpolated into the string in the order
+		// that they are passed into this function from left to right.
+		//
+		// If there are more arguments than there are interpolation indicators `%$`
+		// in the format string, the extra arguments are ignored.
+		template <typename... ParamsT>
+		[[nodiscard]] explicit Format(std::string_view const formatString, ParamsT&&... args)
+			: m_Objects(std::forward<ParamsT>(args)...),
+			  m_FormatString(formatString) {}
+
+		// Convenience method.
+		//
+		// Calls `::Logger::WriteLine(*this)`.
+		void constexpr WriteLine() const &
+		{
+			return ::Logger::WriteLine(*this);
+		}
+
+		// Convenience method.
+		//
+		// Calls `::Logger::WriteLine(std::move(*this))`.
+		void constexpr WriteLine() &&
+		{
+			return ::Logger::WriteLine(std::move(*this));
+		}
+	};
+
+	// Deduction guide.
+	// 
+	// Deduce the type of a `Format` object from the arguments passed to its constructor.
+	template <typename... ParamsT>
+	Format(std::string_view const, ParamsT&&...) -> Format<ParamsT&&...>;
+
+private:
+
+	// Global logger.
+	static Logger ms_Logger;

```

Comment:
Yes, it would be okay to have the singleton be defined outside of the class definition and be used like this: `g_Logger.WriteLine("message")`. Then, most if not all of the class member functions should be made into instance member functions.

However, with the current implementation, none of the instance member functions of `Logger` are thread-safe. You mentioned in another comment about making the mutex an instance member, which would imply that the logger instances are innately thread-safe, which would fix the issue. You also need to decide if you would like `Logger::Write` and `Logger::FormatWrite` to remain `public`.